homeassistant/components/recorder/__init__.py

```diff
@@ -313,24 +308,41 @@ def _purge_old_data(self):
 
         purge_before = dt_util.utcnow() - timedelta(days=self.purge_days)
 
+        session = Session()
+
         _LOGGER.info("Purging events created before %s", purge_before)
-        deleted_rows = Session().query(Events).filter(
+        deleted_rows = session.query(Events).filter(
             (Events.created < purge_before)).delete(synchronize_session=False)
         _LOGGER.debug("Deleted %s events", deleted_rows)
 
         _LOGGER.info("Purging states created before %s", purge_before)
-        deleted_rows = Session().query(States).filter(
+        deleted_rows = session.query(States).filter(
             (States.created < purge_before)).delete(synchronize_session=False)
         _LOGGER.debug("Deleted %s states", deleted_rows)
 
-        Session().commit()
-        Session().expire_all()
+        self._commit(session)
+        session.expire_all()
 
         # Execute sqlite vacuum command to free up space on disk
         if self.engine.driver == 'sqlite':
             _LOGGER.info("Vacuuming SQLite to free space")
             self.engine.execute("VACUUM")
 
+    @staticmethod
+    def _commit(session):
+        """Commit a session."""
+        import sqlalchemy.exc
+        for _ in range(0, RETRIES):
+            try:
+                session.commit()
+                return
+            except sqlalchemy.exc.OperationalError as e:
+                log_error(e, retry_wait=QUERY_RETRY_WAIT,
+                          rollback=True)
```

Comment:
In that case this needs some more thought. I thought it might be related to network error or something in the commit process. 
In what cases will repeating the same work give a better result?